src/main.rs

```diff
@@ -38,31 +38,36 @@ struct MarketData {
 
 impl MarketData {
     fn insert_and_validate(&mut self, hash: &str, filerequest: &FileRequest) {
```

Comment:
Suggestion: Sorry that I didn't mention this earlier, but it would be fine to actually take the ownership of the `hash` as well since we're not using the actual value afterwards inside `register_file`. With taking the ownership of the value now, we can avoid the unnecessary copying with the `to_string` method now.
# `register_file`
```Rust
async fn register_file(
        &self,
        request: Request<RegisterFileRequest>,
    ) -> Result<Response<()>, Status> {
      let file_hash = file_request.file_hash.clone();
      .
      .
      .
      market_data.insert_and_validate(file_hash, file_request);
      .
      .
      .
}
```
# `insert_and_validate`
```Rust
fn insert_and_validate(&mut self, hash: String, filerequest: FileRequest) {
    let producers = self.files.entry(hash).or_default();
    let current_time = get_current_time();
    .
    .
    .
}